tests/test_download.py

```diff
@@ -76,89 +89,109 @@ def test_download_from_s3_with_endpoint_URL(self, remote_local_file: Any):
             mock_remote_filepath, _ = remote_local_file(cloud_prefix='s3://', filename=file_name)
             client = boto3.client('s3', region_name='us-east-1')
             client.put_object(Bucket=MY_BUCKET, Key=os.path.join(MY_PREFIX, file_name), Body='')
-            download_from_s3(mock_remote_filepath, tmp.name, 60)
+            downloader = S3Downloader()
+            downloader.download(mock_remote_filepath, tmp.name, 60.0)
             assert os.path.isfile(tmp.name)
             assert os.environ['S3_ENDPOINT_URL'] == R2_URL
 
     @pytest.mark.usefixtures('s3_client', 's3_test', 'remote_local_file')
     def test_clienterror_exception(self, remote_local_file: Any):
         with pytest.raises(ClientError):
             mock_remote_filepath, mock_local_filepath = remote_local_file(cloud_prefix='s3://')
-            download_from_s3(mock_remote_filepath, mock_local_filepath, 60)
+            downloader = S3Downloader()
+            downloader.download(mock_remote_filepath, mock_local_filepath, 60)
 
     @pytest.mark.usefixtures('s3_client', 's3_test', 'remote_local_file')
     def test_invalid_cloud_prefix(self, remote_local_file: Any):
         with pytest.raises(ValueError):
             mock_remote_filepath, mock_local_filepath = remote_local_file(cloud_prefix='s9://')
-            download_from_s3(mock_remote_filepath, mock_local_filepath, 60)
+            downloader = S3Downloader()
+            downloader.download(mock_remote_filepath, mock_local_filepath, 60)
 
 
 class TestGCSClient:
 
     @pytest.mark.usefixtures('gcs_hmac_client', 'gcs_test', 'remote_local_file')
     def test_download_from_gcs(self, remote_local_file: Any):
-        with tempfile.NamedTemporaryFile(delete=True, suffix='.txt') as tmp:
-            file_name = tmp.name.split(os.sep)[-1]
+        with tempfile.TemporaryDirectory() as tmp_dir:
```

Comment:
Wondering, why change from `with tempfile.NamedTemporaryFile(delete=True, suffix='.txt') as tmp:`, is this a better approach?

Reply:
I guess 2 parts

1. I moved the logic of `if os.path.exists(local): return` to `CloudDownloader.download()` which means these tests would fail as the files themselves already existed. In the past, the check of the local file only happens at the function `download_file` but not in the branching implementations of `download_from_s3` for examples.
2. I think in actual usage scenarios, we wouldn't expect that file to exist in the first place yet just by providing the local path to the downloader. So the test now would more realistically reflect the situation